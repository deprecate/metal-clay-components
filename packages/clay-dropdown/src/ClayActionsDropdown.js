import 'clay-icon';
import './ClayDropdownBase';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import itemsValidator from './items_validator';
import templates from './ClayActionsDropdown.soy.js';

/**
 * Implementation for Metal Clay Action Dropdown.
 */
class ClayActionsDropdown extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayActionsDropdown.STATE = {
	/**
	 * Button configuration to place a button at dropdown footer.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	button: Config.shapeOf({
		label: Config.string().required(),
		style: Config.oneOf(['primary', 'secondary']).value('primary'),
		type: Config.oneOf(['button', 'reset', 'submit']),
	}),

	/**
	 * Caption text of the dropdown.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	caption: Config.string(),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Flag to indicate if menu is expanded.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {?bool}
	 * @default false
	 */
	expanded: Config.bool().value(false),

	/**
	 * Help text to be shown on top of the open dropdown.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	helpText: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * List of menu items.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {!Array}
	 * @default undefined
	 */
	items: itemsValidator.required(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayActionsDropdown
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),
};

defineWebComponent('clay-actions-dropdown', ClayActionsDropdown);

Soy.register(ClayActionsDropdown, templates);

export {ClayActionsDropdown};
export default ClayActionsDropdown;
