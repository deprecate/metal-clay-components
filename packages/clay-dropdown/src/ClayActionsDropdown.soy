{namespace ClayActionsDropdown}

/**
 * This renders an actions dropdown.
 */
{template .render}
	{@param items: list<?>}
	{@param spritemap: string}
	{@param? button: ?}
	{@param? caption: string}
	{@param? elementClasses: string}
	{@param? expanded: bool}
	{@param? handleButtonClick_: any}
	{@param? helpText: string}
	{@param? id: string}
	{@param? itemsIconAlignment: string}

  {let $dropdownElementClasses kind="text"}
    dropdown-action
    {if $elementClasses}
      {sp}{$elementClasses}
    {/if}
  {/let}

  {let $trigger kind="html"}
    {call ClayIcon.render}
      {param spritemap: $spritemap /}
      {param symbol: 'ellipsis-v' /}
    {/call}
  {/let}

	{call ClayDropdownBase.render}
    {param button: $button /}
    {param caption: $caption /}
    {param elementClasses: $dropdownElementClasses /}
    {param events: ['buttonClicked': $handleButtonClick_] /}
    {param expanded: $expanded /}
    {param helpText: $helpText /}
    {param id: $id /}
    {param items: $items /}
    {param itemsIconAlignment: $itemsIconAlignment /}
    {param label: $trigger /}
    {param ref: 'dropdown' /}
    {param spritemap: $spritemap /}
  {/call}
{/template}
