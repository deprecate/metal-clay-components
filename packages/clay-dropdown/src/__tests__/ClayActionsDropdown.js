import ClayActionsDropdown from '../ClayActionsDropdown';

let clayActionsDropdown;

let items = [
	{
		items: [
			{
				label: 'Edit',
				url: '#1',
			},
			{
				label: 'Preview',
				url: '#1',
			},
		],
		separator: true,
		type: 'group',
	},
	{
		items: [
			{
				label: 'Expire',
				url: '#1',
			},
			{
				label: 'View History',
				url: '#1',
			},
		],
		separator: true,
		type: 'group',
	},
	{
		items: [
			{
				label: 'Permissions',
				url: '#1',
			},
		],
		separator: true,
		type: 'group',
	},
	{
		items: [
			{
				label: 'Copy',
				url: '#1',
			},
			{
				label: 'Move',
				url: '#1',
			},
		],
		separator: true,
		type: 'group',
	},
	{
		items: [
			{
				label: 'Move to the Recycle Bin',
				url: '#1',
			},
		],
		type: 'group',
	},
];

describe('ClayActionsDropdown', function() {
	afterEach(() => {
		if (clayActionsDropdown) {
			clayActionsDropdown.dispose();
		}
	});

	it('should render default markdown', () => {
		clayActionsDropdown = new ClayActionsDropdown({
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayActionsDropdown).toMatchSnapshot();
	});

	it('should render an actions dropdown with classes', () => {
		clayActionsDropdown = new ClayActionsDropdown({
			elementClasses: 'my-custom-class',
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayActionsDropdown).toMatchSnapshot();
	});

	it('should render an actions dropdown with id', () => {
		clayActionsDropdown = new ClayActionsDropdown({
			id: 'myId',
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayActionsDropdown).toMatchSnapshot();
	});

	it('should render an expanded action dropdown', () => {
		clayActionsDropdown = new ClayActionsDropdown({
			expanded: true,
			items: items,
			label: 'Trigger',
			spritemap: 'icons.svg',
		});

		expect(clayActionsDropdown).toMatchSnapshot();
	});
});
