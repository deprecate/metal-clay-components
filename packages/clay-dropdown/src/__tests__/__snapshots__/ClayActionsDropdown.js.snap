// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayActionsDropdown should render an expanded action dropdown 1`] = `
<div class="dropdown show dropdown-action">
  <button aria-expanded="true" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
      <use xlink:href="icons.svg#ellipsis-v"></use>
    </svg>
  </button>
  <div class="dropdown-menu show" style="top: 0px; left: 0px;">
    <div>
      <a class="dropdown-item" href="item1url">Item 1</a>
    </div>
    <div>
      <a class="dropdown-item" href="item2url">Item 2</a>
    </div>
  </div>
</div>
`;

exports[`ClayActionsDropdown should render default markdown 1`] = `
<div class="dropdown dropdown-action">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
      <use xlink:href="icons.svg#ellipsis-v"></use>
    </svg>
  </button>
  <div class="dropdown-menu">
    <div>
      <a class="dropdown-item" href="item1url">Item 1</a>
    </div>
    <div>
      <a class="dropdown-item" href="item2url">Item 2</a>
    </div>
  </div>
</div>
`;
